interface: changed virtual scroll implementation and added to chat

tylershuster opened the pull request:
This PR changes the virtualization component to react-virtuoso, which has the advantage of calculating layout instead of needing the height to be set explicitly.

This makes it possible to virtualize the chat scroll.

The features I'm aware of are at least at parity in this PR, if not better. It no longer uses flow-reverse-column and mostly does away with the idea of pages (see caveats below). I deleted some concepts and renamed others because *I think* they were artifacts of the history of this component.

There is a known issue where (on my ship) messages were out of order with their `number` (as in, I could dbug chat-store and the messages would not have the same number as if I loaded the room on a comet. urbit dialogues specifically) and for whatever reason would load properly in the old implementation but not in the new. I would appreciate some insight there.

I would appreciate a review from @tacryt-socryp, @g-a-v-i-n, @liam-fitzgerald, and/or @matildepark when you get some time. Any comments would be appreciated. I'm putting this in draft because I know it needs some more work to really be polished but I'd like some feedback before moving forward.

matildepark commented:
Wow, big changes here.

>  It no longer uses flow-reverse-column and mostly does away with the idea of pages (see caveats below).

Cheers, will peruse.

liam-fitzgerald commented on pkg/interface/src/views/apps/chat/components/lib/chat-window.tsx:
```diff
@@ -1,25 +1,64 @@
+import GlobalApi from "~/logic/api/global";
+
+// This component gets passed an envelopes prop with 20 envelopes
+// Why? No one knows.
```
this is because the initial chat-view subscription returns 20 envelopes for all chats. More are requested when the chat is selected

tylershuster replied:
Oh ok, so this is an arvo feature not a landscape one

liam-fitzgerald commented on pkg/interface/src/views/apps/chat/components/lib/chat-window.tsx:
```diff
@@ -1,25 +1,64 @@
+}
+
+export class ChatWindow extends Component<ChatWindowProps, ChatWindowState> {
+  private fetchPending: boolean;
```
if you're declaring these members, you can also initialise them here and save you from doing it in the constructor. (It will also infer the type)

tylershuster replied:
Sorry, which are you referring to? Point me in the direction of a component with similar syntax if you don't mind.

liam-fitzgerald replied:
`private fetchPending = false;`

liam-fitzgerald commented on pkg/interface/src/views/apps/chat/components/lib/chat-window.tsx:
```diff
@@ -126,68 +142,104 @@ export class ChatWindow extends Component {
+        ? mailboxSize - unreadCount - 1 // put the one right before at the top
+        : mailboxSize - 1; // Just load it at the end
+    
+    const overscan = 100;
```
should probably be declared outside of the function scope as an all caps constant.

liam-fitzgerald commented on pkg/interface/src/views/apps/chat/components/lib/chat-window.tsx:
```diff
@@ -126,68 +142,104 @@ export class ChatWindow extends Component {
-          }
-        </ChatScrollContainer>
+        <BacklogElement isChatLoading={isChatLoading} />
+        <ResubscribeElement {...{ api, host: ship, station, isChatUnsynced}} />
```
woah I didn't realise you could do this

tylershuster replied:
Yeah, it's the same feature as 
const { props } = this

liam-fitzgerald reviewed (approved):
Thanks for this! Looks good to me, but a review from @tacryt-socryp would be nice as he just refactored some of this.

matildepark commented:
Does this load the entire room? UC General was blank in testing for quite a while ... before it suddenly jumped 1,300 messages up. The entire scrollback does seem to be there...

matildepark commented:
It's pretty nice. The cutoff for unloading seems to be like 75% of the actual chat window height, weirdly. And yes, we need @tacryt-socryp to scrutinise this...

tylershuster commented:
@matildepark It doesn't load the entire room. Basically it gets given the initial 20, fetches the `DEFAULT_BACKLOG_SIZE`, and then when you scroll up there's a few things that happen. One, if you're within `DEFAULT_BACKLOG_SIZE` of the top OR by some twist of fate you've actually *reached* the top, it loads `DEFEAULT_BACKLOG_SIZE` messages.

Again, there is some weirdness with message ordering but it seems to be a glitch in my ship. If anyone knows, I've asked tacryt about using the `number` property of the message once `graph-store` launches, because that will make this bulletproof.

matildepark commented:
pong @tacryt-socryp 

(Also conflict @tylershuster)

tacryt-socryp commented:
Strangely enough, in Safari, when you scroll up using the mouse, the entire page doesn't render at all.
<img width="757" alt="Screen Shot 2020-09-02 at 1 31 45 PM" src="https://user-images.githubusercontent.com/1377557/92022308-adde8800-ed20-11ea-89fa-01e963633c6a.png">

tacryt-socryp commented:
Another strange artifact where the page partially renders:
<img width="757" alt="Screen Shot 2020-09-02 at 1 32 59 PM" src="https://user-images.githubusercontent.com/1377557/92022426-dbc3cc80-ed20-11ea-9014-16ce903cbee9.png">

tacryt-socryp commented:
A bit strange. I'm unsure of whether this is polished enough to go out, though it seems due to the library's behavior, not the implementation. This seems like a product decision to me, not a technical one. The code style is good, but if it does go in, please remove the comments.

tacryt-socryp commented:
@urcades @matildepark Up to you.

tylershuster commented:
I agree it's not ready to go out. I will check it mor thoroughly in other browsers. If you think the code style is acceptable I will probably just make small tweaks to get it to work.

matildepark commented:
Ping when it's more polished and we'll take another look.

urcades commented:
I'll also provide feedback when item is updated.

tylershuster commented:
Ok, turns out I was fetching messages incorrectly. This seems to work very nicely. Check out the [video here](https://marpem-files.sfo2.digitaloceanspaces.com/Scrolling.mov).

@urcades you can see some of the behavior I mentioned here. The library allows for scroll placeholders, and I took a rough stab at them here, drawing sigils every five messages and pseudo-randomly setting the length of the message elements. It's a new design pattern so I would appreciate your input.

I would also appreciate some insight on how you would prefer scroll slamming and idleness to works. It can now load and render at an arbitrary point in the list (I tested it with 300+ unread messages and it always loaded at the correct point), but I don't know expected/desired behavior around what to do when a new message is added.

There's a little to clean up and some conflicts to resolve, but I'm much happier with the way this works.

matildepark commented:
@tylershuster @urcades is out for a week (but I've seen him post around anyway) so I'm happy to handle in his absence.

> I would also appreciate some insight on how you would prefer scroll slamming and idleness to works. It can now load and render at an arbitrary point in the list (I tested it with 300+ unread messages and it always loaded at the correct point), but I don't know expected/desired behavior around what to do when a new message is added.

See [previous](https://github.com/urbit/urbit/pull/2825#issuecomment-621296359) [conversations](https://github.com/urbit/urbit/issues/2864#issuecomment-624736105) on this matter (since May?...). Basically we only want to pin us to the bottom if we're already at / near the bottom (1-2 messages, 10% of the viewport, whatever), or if we send a message.

Will test...

matildepark commented:
> @matildepark It doesn't load the entire room. Basically it gets given the initial 20, fetches the `DEFAULT_BACKLOG_SIZE`, and then when you scroll up there's a few things that happen. One, if you're within `DEFAULT_BACKLOG_SIZE` of the top OR by some twist of fate you've actually _reached_ the top, it loads `DEFEAULT_BACKLOG_SIZE` messages.
> 
> Again, there is some weirdness with message ordering but it seems to be a glitch in my ship. If anyone knows, I've asked tacryt about using the `number` property of the message once `graph-store` launches, because that will make this bulletproof.

Are you *sure*? In Safari you're sitting on a blank screen for about ~30 seconds for General on UC, the scrollbar is *long* and you can scroll all the way to April 7 without any wait.

matildepark commented:
This screen blanking seems a lil aggressive also, it makes me lose track of where I was when I scrolled up. Seemingly fires at the smallest scroll up. 
![Sep-06-2020 19-01-50](https://user-images.githubusercontent.com/20846414/92336899-81ee3a00-f073-11ea-8cf4-2d2fae90228e.gif)

tylershuster commented:
@matildepark regarding the initial load: Loading a room I already have been in before loads instantly. The scrollbar is long as a *feature* — it knows the `mailboxSize` so it guesses how long based on the elements it has. It updates when it finds them. I just subscribed to a new room and it took forever to join, but the `fetchMessages` method didn't even fire. Then I loaded it in a new browser (checked Safari too) and stuff was fast. Can you give me some more details? If you use React DevTools you can inspect what a given component has. Scrolling really is wicked fast with this implementation.

As for the scrolling, I agree and that should be easy to change.

Thanks for the feedback on slamming. I think that's pretty easy to implement

matildepark commented:
Idk, it's simple as "I open Safari to localhost:9000 and click General". [Vid](https://drive.google.com/file/d/1c28_rlAemratIwk2DtycMf3tGdBMv4Fq/view?usp=sharing). We just want it to show the latest 20 instead of making us wait for the whole thing. Idk what else it's doing; rooms with less history take less time.

matildepark commented:
Also want to note that we seemingly never fire the 'read' action when we hit bottom. Even clicking 'mark as read' does nothing?

tylershuster commented:
Oh f that is debugging code.

tylershuster commented:
I'll have to investigate the loading behavior.

tylershuster commented:
You were right — partially because of some bad math on my part, and partially because the reducer didn't dedupe messages after loading, so you could end up with as many copies of the same message as times you had loaded the channel in a session.

Working on slamming behavior now.

tylershuster commented:
Alright, racking up the conflicts but the implementation is much smoother now thanks to a) fixing the way that chat-update reduces messages and b) debouncing fetches so it doesn't try and grab new messages until you stop scrolling

tylershuster commented:
Also added a new feature where you can click the unread notice to scroll to that position

tylershuster commented on pkg/interface/src/logic/reducers/chat-update.ts:
```diff
@@ -49,8 +50,11 @@ export default class ChatReducer<S extends ChatState> {
     if (data) {
-      state.inbox[data.path].envelopes =
-        state.inbox[data.path].envelopes.concat(data.envelopes);
+      state.inbox[data.path].envelopes = _.unionBy(
```
This prevents dupes

tylershuster commented on pkg/interface/src/logic/api/chat.ts:
```diff
@@ -12,7 +12,7 @@ export default class ChatApi extends BaseApi<StoreState> {
    */
   fetchMessages(start: number, end: number, path: Path) {
-    fetch(`/chat-view/paginate/${start}/${end}${path}`)
+    return fetch(`/chat-view/paginate/${start}/${end}${path}`)
```
Allows for post-fetch behavior

tylershuster commented on pkg/interface/src/views/apps/chat/components/lib/chat-message.tsx:
```diff
@@ -2,16 +2,15 @@ import React, { PureComponent, Fragment } from "react";
     previousMsg,
     nextMsg,
-    isLastUnread,
+    isFirstUnread,
```
This is because we render in the correct order now

tylershuster commented on pkg/interface/src/views/components/RemoteContent.tsx:
```diff
@@ -116,7 +116,10 @@ export default class RemoteContent extends Component<RemoteContentProps, RemoteC
       );
     } else if (isOembed && remoteContentPolicy.oembedShown) {
-      this.loadOembed();
+      if (!this.state.embed) {
```
Fixes an uncaught bug where it would just fetch and fetch and fetch

tylershuster commented:
Cleaned up; behaves very nicely now

matildepark commented:
Good stuff, will test.

matildepark commented:
- The placeholder needs a gentler gray, basically, it needs to get out of the way, it flashes quick and it's tolerable if it's barely noticeable
- Chat doesn't seem to keep us pinned at bottom if a new message comes in and we're at the very bottom? It shows the "1 new message" but tbh I wasn't in focus, I have to send the incoming message in another window, so possibly it's working as intended if it's based on focus

Otherwise, yeah a total improvement imo

tylershuster commented:
Hm, good points. I'll clean up the placeholder and account for dark mode after the sun sets. The pinning worked last time I checked; maybe I changed something.

I don't think I changed the padding, but maybe reversing the scroll direction did something.

matildepark commented:
It also does a clean flash and load of messages on navigating to Settings or per room (as it doesnt in livenet). I wonder how to prevent a re-render if we don't need to.

liam-fitzgerald commented:
Just tried this! This is mad nice. I have no qualms with it, other than the flash of message that occurs when i nav to a channel.

tylershuster commented:
Alright, I fixed the padding thing (also fixes #3332). The flash is unavoidable, I think. The one downside of not using the flex-reverse implementation. Though I did reduce its severity.

matildepark reviewed (approved):
I'm content. If anything is left here we can sort out in the next few days.
